src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1508 @@
+#include "utilities/macros.hpp"
+#include <type_traits>
+
+class IntrusiveListEntry;
+class IntrusiveListImpl;
+
+/**
+ * The IntrusiveList class template provides a doubly-linked list in
+ * which the links between elements are embedded directly into objects
+ * contained in the list.  As a result, there are no copies involved
+ * when inserting objects into the list or referencing list objects,
+ * and removing an object from a list need not involve destroying the
+ * object.
+ *
+ * To be used in a IntrusiveList, an object must have a
+ * IntrusiveListEntry member.  An IntrusiveList is associated with the
+ * class of its elements and the entry member.
+ *
+ * An object can be in multiple lists at the same time, so long as
+ * each list uses a different entry member.  That is, the class of the
+ * object must have multiple IntrusiveListEntry members, one for each
+ * list the object is simultaneously an element.
+ *
+ * The API for IntrusiveList is modelled on the containers provided by
+ * the C++ standard library.  In particular, iteration over the
+ * elements is provided via iterator classes.
+ *
+ * IntrusiveLists support polymorphic elements.  Because the objects
+ * in a list are externally managed, rather than being embedded values
+ * in the list, the actual type of such objects may be more specific
+ * than the list's element type.
+ *
+ * * T is the class of the elements in the list.  Must be a possibly
+ * const-qualified class type.
+ *
+ * * entry_member is a pointer to class member referring to the
+ * IntrusiveListEntry subobject of T used by this list.
+ *
+ * * has_size determines whether the list has a size()
+ * operation, returning the number of elements in the list.  If the
+ * operation is present, it has constant-time complexity.  The default
+ * is to not provide a constant-time size() operation.
+ *
+ * * Base is the base class for the list.  This is typically
+ * used to specify the allocation class.  The default is void, indicating
+ * no allocation class for the list.
+ *
+ * The value type for a list may be const-qualified.  Such a list provides
+ * const iterators and access to const-qualified elements.  A list whose value
+ * type is not const-qualified can provide both const and non-const iterators
+ * and can provide access to either const or unqualified elements, depending
+ * on the operation used.  A list that is itself const only provides const
+ * iterators and access to const-qualified elements.  A const object cannot be
+ * added to a list whose value type is not const-qualified, as that would be
+ * an implicit casting away of the const qualifier.
+ *
+ * Some operations that remove elements from a list take a disposer argument.
+ * This is a function or function object that will be called with one
+ * argument, a const reference to a removed element.  This function should
+ * "dispose" of the argument object when called, such as by deleting the
+ * object.  The result of the call is ignored.
```

Comment:
A lot of this information is more easily seen by reading the code. Saying something like:

>Operations that remove elements from a list take a disposer function as an argument. A disposer should free up the memory associated with the object.

Shorter documentation is quicker to read and easier to maintain.

Reply:
There are other uses for disposers than freeing memory.  StefanJ found a nice
one while prototyping a conversion of FreeRegionList (which didn't work until
the recent commit to fix the disposer API).

And again, I don't want to require reading the code to just use it.